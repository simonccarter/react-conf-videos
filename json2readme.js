const fs = require('fs')
const {
  is,
  zip,
  map,
  uniq,
  prop,
  pluck,
  curry,
  ifElse,
  sortBy,
  filter,
  propEq,
  compose,
  reverse,
  forEachObjIndexed
} = require('ramda')

const conferenceVids = JSON.parse(fs.readFileSync('./public/assets/conferenceVids.json'))

const extratYearFromDDMMYY = date => date.split('-')[2]

const computePlaylistDetails = ifElse(
  is(String),
  playlist => ` - [playlist](${playlist})`,
  (playlist) => {
    let playlists = ''
    forEachObjIndexed((val, key) => {
      playlists += ` - [playlist: ${key}](${val})`
    }, playlist)
    return playlists
  }
)

const conferenceDetails = ({ title, website, playlist = '' }) => {
  let output = ''
  output += `\n### ${title}\n\n##### ${website ? `[website](${website})` : ''}`
  output += computePlaylistDetails(playlist)
  return `${output}\n`
}

const tableHeader = () => '\n| Title/Link        | Presenter  | Length |\n| ------------- |:-------------:| -----:|'

const conferenceVideos = (videos) => {
  const createTable = curry((title, videos) => {
    let output = title ? `\n#### ${title}\n` : '\n'
    output += tableHeader()
    output += videos.reduce((acc, {
      title, link, presenter: { name }, length
    }) => {
      acc += `\n[${title}](${link}) | ${name} | ${length}`
      return acc
    }, '')
    output += '\n'
    return output
  })

  const splits = compose(uniq, pluck('split'))(videos)
  let output = ''
  if (!splits.length) {
    output = createTable('', videos)
  } else {
    output = map(split => compose(
      createTable(split),
      filter(propEq('split', split)) // extract out matching videos for split
    )(videos)
    )(splits).join('')
  }
  return output
}

const createBody = conferenceVids => conferenceVids.reduce((acc, conference, idx, arr) => {
  // output year if different to last one, or if first conference in list
  const { date } = conference
  const year = extratYearFromDDMMYY(date)
  const pYear = extratYearFromDDMMYY(arr[idx ? idx - 1 : 0].date)
  if (idx === 0 || year !== pYear) {
    acc += `\n\n## ${year}\n`
  }
  // conf details
  acc += conferenceDetails(conference)
  // video table(s)
  acc += conference.videos ? conferenceVideos(conference.videos) : ''
  return acc
}, '')

const createHead = () => '# React.js Conference Videos\nList of react conference videos.\n' +
'[www.reactjsvideos.com](https://www.reactjsvideos.com)'

const createNavLinks = (conferenceVids) => {
  // produce nested arrays containing [[title, year], [title, year]...]
  const titlesAndYears = zip(
    pluck('title', conferenceVids),
    compose(
      map(extratYearFromDDMMYY),
      pluck('date')
    )(conferenceVids)
  )

  // create nav list over them a
  const computeLinks = () => titlesAndYears.reduce((acc, [title, year], idx) => {
    if (idx === 0 || year !== titlesAndYears[idx > 0 ? idx - 1 : 0][1]) {
      acc += `\n* ${year}`
    }
    acc += `\n  * [${title}](#${title.replace(/\s/g, '-').replace(/\.+/g, '').toLowerCase()})`
    return acc
  }, '')

  return `\n## Quick Links\n${computeLinks()}\n * [Contributing](#contributing)`
}

// sort function by date
const sortByDate = sortBy(compose(
  extratYearFromDDMMYY,
  prop('date')
))

// sort vids and reverse so most recent are first within array
const sortJSONByDate = compose(
  reverse,
  sortByDate
)

const createFooter = () => `

## Contributing

To add a conference and it's videos, or to simply fix a typo: 

1. edit \`public/assets/conferenceVids.json\`
2. run \`node json2readme > README.md\`
`

const run = (conferenceVids) => {
  // should already be sorted, but in case not
  const sorted = sortJSONByDate(conferenceVids)

  const head = createHead()
  const navLinks = createNavLinks(sorted)
  const body = createBody(sorted)
  const footer = createFooter()

  return `${head}\n${navLinks}\n${body}\n${footer}`
}

console.log(run(conferenceVids))

