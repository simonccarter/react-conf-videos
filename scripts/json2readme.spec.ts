import fs from 'fs';
import { run } from './json2readme';

describe('json2reamd', () => {
  describe('run', () => {
    it('should match snapshot', () => {
      // arrange
      const inputJSON = JSON.parse(
        fs.readFileSync('./scripts/testInput.json', 'utf8')
      );

      // act
      const result = run(inputJSON);

      // assert
      expect(result).toMatchSnapshot();
    });
  });
});
