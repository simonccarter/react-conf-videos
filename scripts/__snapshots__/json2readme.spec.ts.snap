// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json2reamd run should match snapshot 1`] = `
"# React.js Conference Videos
List of react conference videos.
[www.reactjsvideos.com](https://www.reactjsvideos.com)

## Quick Links

* 2017
  * [React Rally 2017](#react-rally-2017)
  * [React Conf 2017](#react-conf-2017)
* 2016
  * [ReactiveConf 2016](#reactiveconf-2016)
 * [Contributing](#contributing)


## 2017

### React Rally 2017

##### [website](http://www.reactrally.com/) - [playlist](https://www.youtube.com/watch?v=2rhkgB8Cohc&list=PLUqJuYVCz-7eHxPEPz8hJCz2pW6N0H531)

#### Day 1

| Title/Link        | Presenter  | Length |
| ------------- |:-------------:| -----:|
[UNPKG: The CDN for everything on npm](https://www.youtube.com/watch?v=2rhkgB8Cohc&list=PLUqJuYVCz-7eHxPEPz8hJCz2pW6N0H531) | Michael Jackson | 31:46
[D3 & React](https://www.youtube.com/watch?v=ladXdJ3KKd4&index=2&list=PLUqJuYVCz-7eHxPEPz8hJCz2pW6N0H531) | Shirley Wu | 26:57

### React Conf 2017

##### [website](http://conf.reactjs.org/) - [playlist](https://www.youtube.com/playlist?list=PLb0IAmt7-GS3fZ46IGFirdqKTIxlws7e0)

#### Day 1

| Title/Link        | Presenter  | Length |
| ------------- |:-------------:| -----:|
[Keynote Part 1](https://www.youtube.com/watch?v=7HSd1sk07uU&index=1&list=PLb0IAmt7-GS3fZ46IGFirdqKTIxlws7e0) | Tom Occhino | 6:36
[Keynote Part 2 - Incrementally Adopting React Native at Facebook](https://www.youtube.com/watch?v=cSUxHv-kH7w&list=PLb0IAmt7-GS3fZ46IGFirdqKTIxlws7e0&index=2) | Jing Chen | 14:13


## 2016

### ReactiveConf 2016

##### [website](https://reactiveconf.com/2016/) - [playlist](https://www.youtube.com/watch?v=gbATtWPUfN0&list=PLa2ZZ09WYepMCRRGCRPhTYuTCat4TiDlX)

#### Day 2

| Title/Link        | Presenter  | Length |
| ------------- |:-------------:| -----:|
[Writing Sound Asynchronous Code with Coroutines](https://www.youtube.com/watch?v=gZz-1OaMpqg&index=29&list=PLa2ZZ09WYepMCRRGCRPhTYuTCat4TiDlX) | Toma≈° Kulich | 35:43



## Contributing

To add a conference and it's videos, or to simply fix a typo: 

1. edit \`public/assets/conferenceVids.json\`
2. run \`yarn run createReadme\`
"
`;
